### PR TITLE
add make targets to lint the code

### LINKED ISSUE #2818: make it easier to lint the code
When working with Python at work (my projects primary language) it's mandatory that the code be lint free; i.e., pylint produces no diagnostics. This has been extremely useful both because pylint sometimes catches real bugs and because I know if I lint my local repo and it isn't lint free I've introduced lint. For those reasons the fish code should also be lint clean. The first step in doing so is to make it easy for people who aren't using something like the OS X Xcode IDE to lint the code. To achieve that I propose adding `lint` and `lint-all` *make* targets. The next step, which we'll tackle in a separate issue, is to deal with all the existing lint diagnostics.

I started work today on a script that can be invoked from the *Makefile* that will run the code through cppcheck and oclint. The `make lint` command will lint just those files being changed that are not yet committed. If there are none then it will lint the files in the most recently committed change. The `make lint-all` command will, as the name implies, lint all the source in the *src* directory.

### PR DESCRIPTION
**Pull request by krader1961**
Fixes #2818

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,15 +1,50 @@
...
+
+To install the lint checkers on Mac OS X using HomeBrew:
+
+```
+brew tap oclint/formulae
+brew install oclint
+brew install cppcheck
+```
+
+To install the lint checkers on Linux distros that use Aot:
```
Comment (faho) [inline CONTRIBUTING.md]: Typo: apt?

File: build_tools/lint.fish [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    end
+end
+if test (uname -m) = "x86_64"
+    set cppcheck_args -D__x86_64__ -D__LP64__ $cppcheck_args
+end
+
+if test $all = no
+    # We haven't been asked to lint all the source. If there are uncommitted
+    # changes lint those, else lint the files in the most recent commit.
+    set pending (git status --short --untracked-files=all | sed -e 's/^ *//')
```
Comment (faho) [inline build_tools/lint.fish]: What's the sed for? With my git, it seems to do nothing. If the intention is to strip whitespace, you could also do `string trim`. Alternatively, it's probably more stable to use `git status --porcelain` (which is guaranteed stable and ignores a few configuration options) and strip everything up to and including the first space.

File: build_tools/lint.fish [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        for arg in $pending
+            set files $files (string split -m 1 ' ' $arg)[2]
+        end
+    else
+        # No pending changes so lint the files in the most recent commit.
+        set files (git show --name-only --pretty=oneline head | tail --lines=+2)
+    end
+
+    # Filter out the non-C/C++ files.
+    for file in $files
```
Comment (faho) [inline build_tools/lint.fish]: `string match -r '.*.c(pp)? -- $files`?

File: build_tools/lint.fish [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            set c_files $c_files $file
+        end
+    end
+else
+    set c_files src/*.cpp
+end
+
+# We now have a list of files to check so run the linters.
+if set -q c_files[1]
+    if command -s cppcheck > /dev/null
```
Comment (faho) [inline build_tools/lint.fish]: You might want `type -q cppcheck` here.

File: CONTRIBUTING.md (hunk 1 above)
Comment (krader1961) [inline CONTRIBUTING.md]: Fixed.

File: build_tools/lint.fish (hunk 2 above)
Comment (krader1961) [inline build_tools/lint.fish]: Yep, it's to strip the leading whitespace from lines like this: ` M CONTRIBUTING.md`. Without trimming that whitespace the string split below doesn't work as expected. I thought about string trim but that seemed less clear in this situation. Thanks for the suggestion to add `--porcelain`. Done.

File: build_tools/lint.fish (hunk 3 above)
Comment (krader1961) [inline build_tools/lint.fish]: Thanks. Done.

File: build_tools/lint.fish (hunk 4 above)
Comment (krader1961) [inline build_tools/lint.fish]: Hahaha! I made the same suggestion a few weeks ago on someone else's change (zanchey's?) and they pointed out the `type` is a function that calls `command` so better to just use the latter. Personally I think the `type -q` formulation is clearer and the extra overhead is probably not even measurable in most scripts. Done.

**Comment by krader1961** [comment]
Merged as 6fa09e6a70d4116e22df36d33e2f7533cbcbe098. Now to do the same thing for fixing style violations.

**Pull request closed without merging**